### PR TITLE
New Stacked Borrows, now with better support for interior mutability

### PR DESCRIPTION
**Pull request by RalfJung**
Waits on https://github.com/rust-lang/rust/pull/55549 to land

### CODE REVIEW

**Comment by RalfJung** [comment]
The rustc PR landed, this is ready for review.

File: src/stacked_borrows.rs [hunk 1]
```diff
@@ -156,124 +139,125 @@ pub struct Stacks {
...
+                }
+                _ => {},
+            }
+        }
+        // See if we can find this borrow.
+        for (idx, &itm) in self.borrows.iter().rev().enumerate() {
+            // Check borrow and stack item for compatibility.
+            match (itm, bor) {
+                (BorStackItem::FnBarrier(_), _) => {
+                    return Err(format!("Trying to reactivate a borrow ({:?}) that lives \
```
Comment (oli-obk) [inline src/stacked_borrows.rs]: This has no tests that I could find

Comment (RalfJung) [reply src/stacked_borrows.rs]: Barriers are not implemented yet.

File: src/lib.rs [hunk 2]
```diff
@@ -476,38 +476,38 @@ impl<'a, 'mir, 'tcx> Machine<'a, 'mir, 'tcx> for Evaluator<'tcx> {
...
         if !ecx.machine.validate || size == Size::ZERO {
             // No tracking
-            Ok(place)
+            Ok(place.ptr)
         } else {
             let ptr = place.ptr.to_ptr()?;
-            let tag = ecx.tag_dereference(ptr, ty, size, mutability.into())?;
-            let ptr = Scalar::Ptr(Pointer::new_with_tag(ptr.alloc_id, ptr.offset, tag));
-            Ok(MemPlace { ptr, ..place })
+            let tag = ecx.tag_dereference(place, size, mutability.into())?;
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/lib.rs]: `tag_reference` is gone anyway in the next version of this branch.

File: src/stacked_borrows.rs (hunk 1 above)
Comment (oli-obk) [reply src/stacked_borrows.rs]: :laughing: ok

**Pull request merged**